plugins/repository-s3/src/main/plugin-metadata/plugin-security.policy

```diff
@@ -36,6 +36,11 @@ grant {
   // See https://github.com/aws/aws-sdk-java/issues/766
   permission java.lang.reflect.ReflectPermission "suppressAccessChecks";
 
+  // AWS's SDK loads the web identity token from this path,
+  // which is where the token is mounted in the Kubernetes
+  // pod's containers by default
+  permission java.io.FilePermission "${/}var${/}run${/}secrets${/}eks.amazonaws.com${/}serviceaccount${/}token", "read";
```

Comment:
Hard-coding this seems like the sort of thing that will bite us in the future. The file is at the location defined by the environment variable `AWS_WEB_IDENTITY_TOKEN_FILE` which today happens to live at that path, but I don't see docs indicating that the literal path can be relied on. Can we instead do something like set up a symlink in the config directory, based on the environment variable.

Reply:
I'm with you and I also don't like this hard-coded path, but I'm not sure there would be major gains in treating this. I believe this implementation is simpler and [AWS itself hard-coded this path on their services](https://github.com/aws/amazon-eks-pod-identity-webhook/blob/599cfe7cfc8c72d48a8870e675e19a8169b3dafe/pkg/handler/handler.go#L115). Changing it would not only be a huge architectural change on their side but also a major breaking one, specially with their own SDKs. Also, setting up the symlink would just be moving this hard-coded path from one place to another, and possibly would make things a little bit trickier to implement (and would involve hard-coding another path - the symlink's). Sorry if I missed your point, but I'm not sure what are the advantages of not doing it this way.

Reply:
The code to which you link isn't a SDK, it appears to be for setting up the environment in which a SDK runs, and in particular it sets the `AWS_WEB_IDENTITY_TOKEN_FILE` environment variable to tell the container where to find the token file so that this path can be changed without breaking any well-behaved SDK clients in the future.

The point of using a symlink here is that we don't need special permissions to read from files within the config directory, and those permissions extend to the targets of symlinks. You are correct that this just moves the problem to a different place, but that different place has usability advantages: a dangling symlink is a lot easier to diagnose than missing permissions, and is something that I'd be more comfortable asking users to address themselves.

Reply:
I'm pretty sure we'll be able to come up with a solution for this somehow, although working out the details might involve some changes to how Elasticsearch is packaged or how it starts up. We can return to this later, the more important thing at the moment is the lack of YAML REST tests for this new functionality

Reply:
Cool! I've been quite busy these last few weeks at work and wasn't able to pay much attention here - but I'll be on my xmas/new year's break soon and I'll add the missing tests asap. I'll also think about this symlink part and see what I can come up with. Thanks for the feedback on this!

Reply:
There are two problems with the proposed solution here. First, the hardcoded file as mentioned will not work long term. Second, we are actually moving towards limiting the ability of plugins to read files outside of their configuration directory.

I have the following proposal to move this forward. A user of the repository-s3 plugin would need to add a symlink to their configuration directory pointing to the `AWS_WEB_IDENTITY_TOKEN_FILE` location. The plugin's policy file would then ask for the `read,readlink` permissions on that path. Eventually we will automatically grant this permission, but in the meantime you can use the `es.path.conf` system property within the policy file to reference the configuration directory. So the permission line would look something like:

```
permission java.io.FilePermission "${es.path.conf}/repository-s3/aws-web-identity-token-file" "read,readlink"  
```

And configuration documentation for the plugin would note the need to add the symlink when launching the node:
```
mkdir -p /etc/elasticsearch/repository-s3
ln -s $AWS_WEB_IDENTITY_TOKEN_FILE /etc/elasticsearch/repository-s3/aws-web-identity-token-file
```